### PR TITLE
Task 1 - 5

### PR DESCRIPTION
**Pull request by VitalyZenkou**

### CODE REVIEW

File: Lecture2/src/task1/Task1App.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
+package task1;
+
+import java.util.Scanner;
+
+public class Task1App {
+
+    private final static String SEPARATOR = "----------------------------------";
```
Comment (nemec8g) [inline Lecture2/src/task1/Task1App.java]: `private static final ` - right order (not final static)

File: Lecture2/src/task1/Task1App.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    private final static int START_INDEX = 5;
+    private final static int END_INDEX = 10;
+    private final static char NEW_SYMBOL = '*';
+    private static String sentence;
+    private static Scanner scanner = new Scanner(System.in);
+
+
```
Comment (nemec8g) [inline Lecture2/src/task1/Task1App.java]: 2 UEL

File: Lecture2/src/task1/Task1App.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        printString(SEPARATOR);
+    }
+
+    private static void printReplace(String sentence) {
+        String replace = sentence.replace(' ', NEW_SYMBOL);
+        printString(replace);
+    }
+
+    private static void printUpper(String sentence) {
+        String upper = sentence.toUpperCase();
```
Comment (nemec8g) [inline Lecture2/src/task1/Task1App.java]: please name variables in such way that you may understand it in few days
like sentenceInUpperCase

File: Lecture2/src/task1/Task1App.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.Scanner;
+
+public class Task1App {
+
+    private final static String SEPARATOR = "----------------------------------";
+    private final static int START_INDEX = 5;
+    private final static int END_INDEX = 10;
+    private final static char NEW_SYMBOL = '*';
+    private static String sentence;
+    private static Scanner scanner = new Scanner(System.in);
```
Comment (nemec8g) [inline Lecture2/src/task1/Task1App.java]: final

File: Lecture2/src/task1/Task1App.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        printReverse(sentence);
+        printAllWords(sentence);
+        printReplace(sentence);
+        printUpper(sentence);
+        printSubstring(sentence);
+        scanner.close();
+    }
+
+    private static void printReverse(String sentence) {
+        StringBuilder buff = new StringBuilder();
```
Comment (nemec8g) [inline Lecture2/src/task1/Task1App.java]: why buff?

File: Lecture2/src/task1/Task1App.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        buff.append(sentence);
+        printString(buff.reverse().toString());
```
Comment (nemec8g) [inline Lecture2/src/task1/Task1App.java]: please try to use simple cycle for it

File: Lecture2/src/task1/Task1App.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    private final static int END_INDEX = 10;
+    private final static char NEW_SYMBOL = '*';
+    private static String sentence;
+    private static Scanner scanner = new Scanner(System.in);
+
+
+    public static void main(String[] args) {
+
+        printString("Enter the sentence: ");
+        sentence = scanner.nextLine();
```
Comment (nemec8g) [inline Lecture2/src/task1/Task1App.java]: inline .nextLine().trim()

File: Lecture2/src/task2/Task2App.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            case "7": {
+                System.out.println("Sunday");
+                break;
+            }
+        }
+        scanner.close();
+    }
+
+    private static boolean isWeekDay(String string) {
+        Pattern p = Pattern.compile("^[1-7]{1,1}$");
```
Comment (nemec8g) [inline Lecture2/src/task2/Task2App.java]: Pattern  - to static final Pattern
in this case you don't need to create new Pattern each time

File: Lecture2/src/Task3/Task3App.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            }
+            sideSize = scanner.nextInt();
+            if (sideSize > 1) {
+                break;
+            } else {
+                System.out.println("Side size must be more than 2!");
+            }
+        } while (true);
+
+        initVar();
```
Comment (nemec8g) [inline Lecture2/src/Task3/Task3App.java]: not clear what `initVar ` method  responsible for

File: Lecture2/src/Task3/Task3App.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        printTop();
+        printBot();
```
Comment (nemec8g) [inline Lecture2/src/Task3/Task3App.java]: printBottom

File: Lecture2/src/Task3/Task3App.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        for (int i = 0; i < sideSize - ONE; i++) {
+            buff1 += ONE;
+            buff2 -= ONE;
+            printLine(buff1, buff2);
+            System.out.println();
+        }
+    }
+
+    private static void printLine(int first, int second) {
+        for (int j = 0; j < maxLength; j++) {
```
Comment (nemec8g) [inline Lecture2/src/Task3/Task3App.java]: usually `j` used in arrays like `array[i][j]`
use `i` if you have simple array

File: Lecture2/src/Task3/Task3App.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    private static void initVar() {
+        maxLength = sideSize * TWO + ONE;
+        buff1 = buff2 = sideSize;
+    }
+
+    private static void printTop() {
+        for (int i = 0; i < sideSize; i++) {
+            if (i == 0) {
```
Comment (nemec8g) [inline Lecture2/src/Task3/Task3App.java]: now you check condition   `if (i == 0)` each iteration in `for` cycle.
move it on top of this method and iterate from `1`

File: Lecture2/src/Task3/Task3App.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+                break;
+            } else {
+                System.out.println("Side size must be more than 2!");
+            }
+        } while (true);
+
+        initVar();
+        printTop();
+        printBot();
+        scanner.close();
```
Comment (nemec8g) [inline Lecture2/src/Task3/Task3App.java]: close scanner immediately when you finish work with it

File: Lecture2/src/task4/model/MarathonStatistic.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
+package task4.model;
+
+public final class MarathonStatistic {
+
+    private static String[] names = { "Elena", "Thomas", "Hamilton", "Suzie", "Phil", "Matt", "Alex", "Emma", "John",
```
Comment (nemec8g) [inline Lecture2/src/task4/model/MarathonStatistic.java]: final

File: Lecture2/src/task4/model/MarathonStatistic.java [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+            "James", "Jane", "Emily", "Daniel", "Neda", "Aaron", "Kate" };
+    private static int[] time = {341, 273, 278, 329, 445, 402, 388, 275, 243, 334, 412, 393, 299, 343, 317, 265};
```
Comment (nemec8g) [inline Lecture2/src/task4/model/MarathonStatistic.java]: final

File: Lecture2/src/task4/model/MarathonStatistic.java [hunk 16]
```diff
@@ -0,0 +1,18 @@
+package task4.model;
+
+public final class MarathonStatistic {
+
```
Comment (nemec8g) [inline Lecture2/src/task4/model/MarathonStatistic.java]: private construnctor

File: Lecture2/src/task4/util/MarathonHelper.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+        int length = 0;
+        List<MarathonParticipant> participants = new ArrayList<>();
+        if (names.length != time.length) {
+            System.out.println("The name array doesn't equals time array!!\n" +
+                    "The length will be equals lover value!");
+            if (names.length > time.length) {
+                length = time.length;
+            }
+        } else {
+            length = names.length;
```
Comment (nemec8g) [inline Lecture2/src/task4/util/MarathonHelper.java]: remove `else` statement using default initialization int length =names.length

File: Lecture2/src/task4/util/MarathonHelper.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+    public static void printParticipants(List<MarathonParticipant> participants) {
+        int placeCounter = 1;
+        for (MarathonParticipant participant : participants) {
+            System.out.printf("Place [%-2d] - %s\n", placeCounter, participant);
+            placeCounter++;
+        }
+    }
+
+    public static void printParticipantByPlace(List<MarathonParticipant> participants, int index) {
```
Comment (nemec8g) [inline Lecture2/src/task4/util/MarathonHelper.java]: method printParticipantByPlace - expect place as param  - not index

File: Lecture2/src/task4/view/Menu.java [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.util.List;
+import java.util.Scanner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Menu {
+
+    private static Scanner scanner = new Scanner(System.in);
+    private String menu;
```
Comment (nemec8g) [inline Lecture2/src/task4/view/Menu.java]: var menu - used only in method printMenu - so no need to move it to class level

File: Lecture2/src/task4/view/Menu.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+            place = scanner.nextInt();
+            if (place < maxPlace) {
+                return place;
+            } else {
+                System.out.println("You entered wrong place, max place is: " + maxPlace);
+            }
+        } while (true);
+    }
+
+    private void printSeparator() {
```
Comment (nemec8g) [inline Lecture2/src/task4/view/Menu.java]: used only once?

File: Lecture2/src/task4/view/Menu.java [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+                }
+            }
+        } while (!menu.equals("q"));
+    }
+
+    private void printByPlace(List<MarathonParticipant> participants) {
+        MarathonHelper.printParticipantByPlace(participants, readPlace(participants.size()));
+    }
+
+    private static boolean isMenu(String string) {
```
Comment (nemec8g) [inline Lecture2/src/task4/view/Menu.java]: isMenu should be after readMenu method

File: Lecture2/src/task4/view/Menu.java [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+        Pattern p = Pattern.compile("^[a-z0-9]{1,1}$");
```
Comment (nemec8g) [inline Lecture2/src/task4/view/Menu.java]: Patternt to static final

File: Lecture2/src/task4/view/Menu.java [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+    private static int readPlace(int maxPlace) {
+        int place = 0;
+        do {
+            System.out.print("Enter place: ");
+            while (!scanner.hasNextInt()) {
+                scanner.next();
+                System.out.println("You entered not a number!");
+            }
+            place = scanner.nextInt();
+            if (place < maxPlace) {
```
Comment (nemec8g) [inline Lecture2/src/task4/view/Menu.java]: 0 place it is incorrect program behavior

File: Lecture2/src/task4/util/MarathonHelper.java (hunk 18 above)
Comment (nemec8g) [reply Lecture2/src/task4/util/MarathonHelper.java]: more over you should get first Participant (get(0)) when  place is 1

File: Lecture2/src/task4/model/MarathonParticipant.java [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+
+    public int getTime() {
+        return time;
+    }
+
+    public int getPlace() {
+        return place;
+    }
+
+    public void setPlace(int place) {
```
Comment (nemec8g) [inline Lecture2/src/task4/model/MarathonParticipant.java]: you don't use this setter

File: Lecture2/src/task5/Task5App.java [hunk 25]
```diff
@@ -0,0 +1,103 @@
+package task5;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Task5App {
+
+    private final static char SYMBOL = '*';
+    private static Scanner scanner = new Scanner(System.in);
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: final

File: Lecture2/src/task5/Task5App.java [hunk 26]
```diff
@@ -0,0 +1,103 @@
+package task5;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Task5App {
+
+    private final static char SYMBOL = '*';
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: static final

File: Lecture2/src/task5/Task5App.java [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+    private static Scanner scanner = new Scanner(System.in);
+    private static List<String> variants = Arrays.asList("Vitaly Zenkou", "Vlad Bankov", "Vitaly Plisko");
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: make all needded var final

File: Lecture2/src/task5/Task5App.java [hunk 28]
```diff
@@ -0,0 +1,103 @@
...
+            System.out.println("The hidden name is " + result.toString());
+            String str = readString();
+            compare(str);
+            isWin = isWin();
+        } while (!isWin);
+        System.out.println("You win! The hidden name is " + hiddenName);
+    }
+
+    private static void chooseHiddenName() {
+        Random random = new Random();
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: inline random

File: Lecture2/src/task5/Task5App.java [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+        do {
+            System.out.println("The hidden name is " + result.toString());
+            String str = readString();
+            compare(str);
+            isWin = isWin();
+        } while (!isWin);
+        System.out.println("You win! The hidden name is " + hiddenName);
+    }
+
+    private static void chooseHiddenName() {
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: common suggestion - static var which changes in different methods - bad practice 
make chooseHiddenName returns String  - and work with this var in main method (pass it in param to other methods)

File: Lecture2/src/task5/Task5App.java [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+        hiddenName = variants.get(random.nextInt(variants.size()));
+        fillStringBySymbol();
+    }
+
+    private static void fillStringBySymbol() {
+        for (int i = 0; i < hiddenName.length(); i++) {
+            if (hiddenName.charAt(i) != ' ') {
+                result.append(SYMBOL);
+            } else {
+                result.append(' ');
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: return result from method

File: Lecture2/src/task5/Task5App.java [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+            }
+        } else {
+            String str = hiddenName.toLowerCase();
+            if (str.equals(string.toLowerCase())) {
+                result.replace(0,string.length(), string);
+            }
+        }
+    }
+
+    private static boolean wasEntered(String string) {
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: string - bad name

File: Lecture2/src/task5/Task5App.java [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+
+    private static boolean isWin() {
+        if (result.toString().toLowerCase().equals(hiddenName.toLowerCase())) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private static void changeSymbolByLetter(String symbol) {
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: return result

File: Lecture2/src/task5/Task5App.java [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private static boolean isString(String string) {
+        Pattern p = Pattern.compile("^[a-z A-z]{1,20}$");
+        Matcher m = p.matcher(string);
+        return m.matches();
+    }
+
+    private static void compare(String string) {
+        if (string.length() == 1 ? true : false) {
```
Comment (nemec8g) [inline Lecture2/src/task5/Task5App.java]: if (string.length() == 1) - enough

**Review by nemec8g — CHANGES_REQUESTED** [review]

File: Lecture2/src/Task3/Task3App.java (hunk 9 above)
Comment (VitalyZenkou) [reply Lecture2/src/Task3/Task3App.java]: Deleted this method. Variables initialize in the main method now.

File: Lecture2/src/Task3/Task3App.java (hunk 10 above)
Comment (VitalyZenkou) [reply Lecture2/src/Task3/Task3App.java]: fix

File: Lecture2/src/Task3/Task3App.java (hunk 13 above)
Comment (VitalyZenkou) [reply Lecture2/src/Task3/Task3App.java]: fix

File: Lecture2/src/Task3/Task3App.java (hunk 12 above)
Comment (VitalyZenkou) [reply Lecture2/src/Task3/Task3App.java]: fix

File: Lecture2/src/Task3/Task3App.java (hunk 11 above)
Comment (VitalyZenkou) [reply Lecture2/src/Task3/Task3App.java]: fix

File: Lecture2/src/task1/Task1App.java (hunk 1 above)
Comment (VitalyZenkou) [reply Lecture2/src/task1/Task1App.java]: fix

File: Lecture2/src/task1/Task1App.java (hunk 4 above)
Comment (VitalyZenkou) [reply Lecture2/src/task1/Task1App.java]: fix

File: Lecture2/src/task1/Task1App.java (hunk 2 above)
Comment (VitalyZenkou) [reply Lecture2/src/task1/Task1App.java]: fix

File: Lecture2/src/task1/Task1App.java (hunk 7 above)
Comment (VitalyZenkou) [reply Lecture2/src/task1/Task1App.java]: fix

File: Lecture2/src/task1/Task1App.java (hunk 5 above)
Comment (VitalyZenkou) [reply Lecture2/src/task1/Task1App.java]: fix

File: Lecture2/src/task1/Task1App.java (hunk 6 above)
Comment (VitalyZenkou) [reply Lecture2/src/task1/Task1App.java]: Rewrite. fix

File: Lecture2/src/task1/Task1App.java (hunk 3 above)
Comment (VitalyZenkou) [reply Lecture2/src/task1/Task1App.java]: fix

File: Lecture2/src/task2/Task2App.java (hunk 8 above)
Comment (VitalyZenkou) [reply Lecture2/src/task2/Task2App.java]: fix

File: Lecture2/src/task4/model/MarathonParticipant.java (hunk 24 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/model/MarathonParticipant.java]: fix

File: Lecture2/src/task4/model/MarathonStatistic.java (hunk 16 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/model/MarathonStatistic.java]: I have private constructor in this class 9 line.

File: Lecture2/src/task4/model/MarathonStatistic.java (hunk 14 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/model/MarathonStatistic.java]: fix

File: Lecture2/src/task4/model/MarathonStatistic.java (hunk 15 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/model/MarathonStatistic.java]: fix

File: Lecture2/src/task4/util/MarathonHelper.java (hunk 17 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/util/MarathonHelper.java]: fix

File: Lecture2/src/task4/view/Menu.java (hunk 19 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/view/Menu.java]: fix

File: Lecture2/src/task4/view/Menu.java (hunk 21 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/view/Menu.java]: fix

File: Lecture2/src/task4/view/Menu.java (hunk 22 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/view/Menu.java]: fix

File: Lecture2/src/task4/view/Menu.java (hunk 23 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/view/Menu.java]: fix

File: Lecture2/src/task4/view/Menu.java (hunk 20 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/view/Menu.java]: Deleted.

File: Lecture2/src/task5/Task5App.java (hunk 26 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 25 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 27 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 28 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 30 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 33 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 31 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 32 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task5/Task5App.java (hunk 29 above)
Comment (VitalyZenkou) [reply Lecture2/src/task5/Task5App.java]: fix

File: Lecture2/src/task4/util/MarathonHelper.java (hunk 18 above)
Comment (VitalyZenkou) [reply Lecture2/src/task4/util/MarathonHelper.java]: fix, rewrite this methods

File: com.course.zenkou/src/Lecture2/task2/Task2App.java [hunk 34]
```diff
@@ -0,0 +1,61 @@
...
+            dayNumber = scanner.next();
+            if (isWeekDay(dayNumber)) {
+                scanner.close();
+                break;
+            } else {
+                System.out.println("You entered wrong day!");
+            }
+        } while (true);
+
+        switch (dayNumber) {
```
Comment (nemec8g) [inline com.course.zenkou/src/Lecture2/task2/Task2App.java:25]: good solution but you may read about  DateFormatSymbols().getWeekdays() from java 8

File: com.course.zenkou/src/Lecture2/task4/util/MarathonHelper.java [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class MarathonHelper {
+
+    private MarathonHelper() {
+    }
+
+    public static List<MarathonParticipant> createParticipants(String[] names, int[] time) {
```
Comment (nemec8g) [inline com.course.zenkou/src/Lecture2/task4/util/MarathonHelper.java:13]: the good place for this method is ParticipantFactory

File: com.course.zenkou/src/Lecture2/task4/util/MarathonHelper.java [hunk 36]
```diff
@@ -0,0 +1,51 @@
...
+        return participants;
+    }
+
+    public static void printParticipants(List<MarathonParticipant> participants) {
+        for (MarathonParticipant participant : participants) {
+            System.out.println(participant);
+        }
+    }
+
+    public static void setPlaces(List<MarathonParticipant> participants) {
```
Comment (nemec8g) [inline com.course.zenkou/src/Lecture2/task4/util/MarathonHelper.java:35]: the are risk - if you provide not sorted collection 
as solution - you may create method holdСompetition in some Сompetition class, where encapsulate this logic  
(sort and set plase)

File: com.course.zenkou/src/Lecture2/task4/model/MarathonParticipant.java [hunk 37]
```diff
@@ -0,0 +1,54 @@
...
+
+    public int getPlace() {
+        return place;
+    }
+
+    public void setPlace(int place) {
+        if (this.place == 0) {
+            this.place = place;
+        } else {
+            System.out.println("The place was set for this participant!");
```
Comment (nemec8g) [inline com.course.zenkou/src/Lecture2/task4/model/MarathonParticipant.java:32]: what about second competition? 
no need to check existing place here

File: com.course.zenkou/src/Lecture2/task2/Task2App.java (hunk 34 above)
Comment (VitalyZenkou) [reply com.course.zenkou/src/Lecture2/task2/Task2App.java:25]: fix

**Pull request merged**